File: src/types/challenge.rs
//! Challenge-related types

use serde::{Deserialize, Serialize};

/// チャレンジ情報
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChallengeInfo {
    pub id: i64,
    pub user_id: i64,
    pub challenge_type: String,
    pub target_metric: String,
    pub target_value: i32,
    pub current_value: i32,
    pub reward_xp: i32,
    pub start_date: String,
    pub end_date: String,
    pub status: String,
    pub completed_at: Option<String>,
    // Computed fields
    pub progress_percent: f32,
    pub remaining_time_hours: i64,
    pub is_completed: bool,
    pub is_expired: bool,
}

impl ChallengeInfo {
    /// Get display name for challenge type
    pub fn challenge_type_label(&self) -> &'static str {
        match self.challenge_type.as_str() {
            "daily" => "デイリー",
            "weekly" => "ウィークリー",
            _ => "その他",
        }
    }

    /// Get display name for target metric
    pub fn target_metric_label(&self) -> &'static str {
        match self.target_metric.as_str() {
            "commits" => "コミット",
            "prs" => "PR",
            "reviews" => "レビュー",
            "issues" => "Issue",
            _ => "その他",
        }
    }

    /// Get icon for target metric
    pub fn target_metric_icon(&self) -> &'static str {
        match self.target_metric.as_str() {
            "commits" => "📝",
            "prs" => "🔀",
            "reviews" => "👀",
            "issues" => "🐛",
            _ => "🎯",
        }
    }

    /// Get status label
    pub fn status_label(&self) -> &'static str {
        match self.status.as_str() {
            "active" => "進行中",
            "completed" => "達成",
            "failed" => "失敗",
            _ => "不明",
        }
    }

    /// Format remaining time as human-readable string
    pub fn remaining_time_label(&self) -> String {
        if self.remaining_time_hours <= 0 {
            return "終了".to_string();
        }

        let hours = self.remaining_time_hours;
        if hours >= 24 {
            let days = hours / 24;
            format!("残り {}日", days)
        } else {
            format!("残り {}時間", hours)
        }
    }
}

/// チャレンジ作成リクエスト
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateChallengeRequest {
    pub challenge_type: String,
    pub target_metric: String,
    pub target_value: i32,
    pub reward_xp: Option<i32>,
}

/// チャレンジ統計
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ChallengeStats {
    pub total_completed: i32,
    pub consecutive_weekly_completions: i32,
    pub active_count: i32,
}

/// チャレンジタイプの選択肢
pub const CHALLENGE_TYPES: &[(&str, &str)] = &[("daily", "デイリー"), ("weekly", "ウィークリー")];

/// ターゲットメトリクスの選択肢
pub const TARGET_METRICS: &[(&str, &str, &str)] = &[
    ("commits", "コミット", "📝"),
    ("prs", "PR", "🔀"),
    ("reviews", "レビュー", "👀"),
    ("issues", "Issue", "🐛"),
];

File: src/types/challenge.ts
// Challenge-related types

/// チャレンジタイプ
export type ChallengeType = 'daily' | 'weekly';

/// ターゲットメトリクス
export type TargetMetric = 'commits' | 'prs' | 'reviews' | 'issues';

/// チャレンジステータス
export type ChallengeStatus = 'active' | 'completed' | 'failed';

/// チャレンジ情報
export interface ChallengeInfo {
  id: number;
  userId: number;
  challengeType: ChallengeType;
  targetMetric: TargetMetric;
  targetValue: number;
  currentValue: number;
  rewardXp: number;
  startDate: string;
  endDate: string;
  status: ChallengeStatus;
  completedAt: string | null;
  // Computed fields
  progressPercent: number;
  remainingTimeHours: number;
  isCompleted: boolean;
  isExpired: boolean;
}

/// チャレンジ作成リクエスト
export interface CreateChallengeRequest {
  challengeType: ChallengeType;
  targetMetric: TargetMetric;
  targetValue: number;
  rewardXp: number | null;
}

/// チャレンジ統計
export interface ChallengeStats {
  totalCompleted: number;
  consecutiveWeeklyCompletions: number;
  activeCount: number;
}

/// チャレンジタイプの選択肢
export const CHALLENGE_TYPES: [string, string][] = [
  ['daily', 'デイリー'],
  ['weekly', 'ウィークリー'],
];

/// ターゲットメトリクスの選択肢
export const TARGET_METRICS: [string, string, string][] = [
  ['commits', 'コミット', '📝'],
  ['prs', 'PR', '🔀'],
  ['reviews', 'レビュー', '👀'],
  ['issues', 'Issue', '🐛'],
];

/// Get display name for challenge type
export function challengeTypeLabel(challengeType: ChallengeType): string {
  switch (challengeType) {
    case 'daily':
      return 'デイリー';
    case 'weekly':
      return 'ウィークリー';
  }
}

/// Get display name for target metric
export function targetMetricLabel(targetMetric: TargetMetric): string {
  switch (targetMetric) {
    case 'commits':
      return 'コミット';
    case 'prs':
      return 'PR';
    case 'reviews':
      return 'レビュー';
    case 'issues':
      return 'Issue';
  }
}

/// Get icon for target metric
export function targetMetricIcon(targetMetric: TargetMetric): string {
  switch (targetMetric) {
    case 'commits':
      return '📝';
    case 'prs':
      return '🔀';
    case 'reviews':
      return '👀';
    case 'issues':
      return '🐛';
  }
}

/// Get status label
export function statusLabel(status: ChallengeStatus): string {
  switch (status) {
    case 'active':
      return '進行中';
    case 'completed':
      return '達成';
    case 'failed':
      return '失敗';
  }
}

/// Format remaining time as human-readable string
export function remainingTimeLabel(remainingTimeHours: number): string {
  if (remainingTimeHours <= 0) {
    return '終了';
  }

  if (remainingTimeHours >= 24) {
    const days = Math.ceil(remainingTimeHours / 24);
    return `残り ${days}日`;
  } else {
    return `残り ${Math.floor(remainingTimeHours)}時間`;
  }
}

